lib/evse_security/crypto/openssl/openssl_supplier.cpp

```diff
@@ -0,0 +1,557 @@
+                    certificates.push_back(std::make_unique<X509HandleOpenSSL>(xi->x509));
+                    xi->x509 = nullptr;
+                }
+            }
+
+            sk_X509_INFO_pop_free(allcerts, X509_INFO_free);
+        } else {
+            throw CertificateLoadException("Certificate (PEM) parsing error");
+        }
+    } else if (encoding == EncodingFormat::DER) {
+        X509* x509 = d2i_X509_bio(bio.get(), nullptr);
+
+        if (x509) {
+            certificates.push_back(std::make_unique<X509HandleOpenSSL>(x509));
+        } else {
+            throw CertificateLoadException("Certificate (DER) parsing error");
+        }
+    } else {
+        throw CertificateLoadException("Unsupported encoding format");
+    }
+
+    return certificates;
+}
+
+std::string OpenSSLSupplier::x509_to_string(X509Handle* handle) {
+    if (X509* x509 = get(handle)) {
+        BIO_ptr bio_write(BIO_new(BIO_s_mem()));
+
+        int rc = PEM_write_bio_X509(bio_write.get(), x509);
+
+        if (rc == 1) {
+            BUF_MEM* mem = NULL;
+            BIO_get_mem_ptr(bio_write.get(), &mem);
+
+            return std::string(mem->data, mem->length);
+        }
+    }
+
+    return {};
+}
+
+std::string OpenSSLSupplier::x509_get_common_name(X509Handle* handle) {
+    X509* x509 = get(handle);
+
+    if (x509 == nullptr)
+        return {};
+
+    X509_NAME* subject = X509_get_subject_name(x509);
+    int nid = OBJ_txt2nid("CN");
+    int index = X509_NAME_get_index_by_NID(subject, nid, -1);
+
+    if (index == -1) {
+        return {};
+    }
+
+    X509_NAME_ENTRY* entry = X509_NAME_get_entry(subject, index);
+    ASN1_STRING* ca_asn1 = X509_NAME_ENTRY_get_data(entry);
+
+    if (ca_asn1 == nullptr) {
+        return {};
+    }
+
+    const unsigned char* cn_str = ASN1_STRING_get0_data(ca_asn1);
+    std::string common_name(reinterpret_cast<const char*>(cn_str), ASN1_STRING_length(ca_asn1));
+    return common_name;
+}
+
+std::string OpenSSLSupplier::x509_get_issuer_name_hash(X509Handle* handle) {
+    X509* x509 = get(handle);
+
+    if (x509 == nullptr)
+        return {};
+
+    unsigned char md[SHA256_DIGEST_LENGTH];
+    X509_NAME* name = X509_get_issuer_name(x509);
+    X509_NAME_digest(name, EVP_sha256(), md, NULL);
+
+    std::stringstream ss;
+    for (int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
+        ss << std::setw(2) << std::setfill('0') << std::hex << (int)md[i];
+    }
+    return ss.str();
+}
+
+std::string OpenSSLSupplier::x509_get_serial_number(X509Handle* handle) {
+    X509* x509 = get(handle);
+
+    if (x509 == nullptr)
+        return {};
+
+    ASN1_INTEGER* serial_asn1 = X509_get_serialNumber(x509);
+    if (serial_asn1 == nullptr) {
+        ERR_print_errors_fp(stderr);
+        return {};
+    }
+
+    BIGNUM* bn_serial = ASN1_INTEGER_to_BN(serial_asn1, NULL);
+
+    if (bn_serial == nullptr) {
+        ERR_print_errors_fp(stderr);
+        return {};
+    }
+
+    char* hex_serial = BN_bn2hex(bn_serial);
```

Comment:
According to the OpenSSL docs, the result of `BN_bn2hex` should be freed later on with `OPENSSL_free()`. Are we handling this properly here?

See https://www.openssl.org/docs/man1.1.1/man3/BN_bn2hex.html

Reply:
Cocerning this, and the rest of this file: I think we should clone and free any OpenSSL-native data structures as soon as possible to avoid causing memory bugs.

Reply:
Correct, solved. For most other structs, we are using std::unique_pointes (defined in detail/opnssl) that should do that safely.